### ISSUE TITLE
CA1031 should not warn in Main or other method called from global context

### ISSUE THREAD

**Issue by NN---** [dotnet/roslyn-analyzers#4148]
### Analyzer

**Diagnostic ID**: [CA1031](https://docs.microsoft.com/visualstudio/code-quality/ca1031)

### Describe the improvement

This should not produce CA1031
```cs
public static void Main()
{
  try 
  {
    MyProgram();
  }
  catch (Exception e)
  {
    Log.Error(e);
  }
}
```

Optionally this too:

```cs
Task.Run(() =>
{
  try 
  {
    MyProgram();
  }
  catch (Exception e)
  {
    Log.Error(e);
  }
});
```

**Comment by Youssef1313** [comment]
I don't see a reason why Main should be treated in a different way?
Catching general exception types isn't preferred whether in Main or not.
For specific cases where it does make sense for the developer to catch a general exception type, the developer should just suppress the warning.

**Comment by NN---** [comment]
If it is not expects in Main, then how should I log exceptions in my program?

**Comment by mavasani** [comment]
Agree with @Youssef1313. The rule is a generic code quality suggestion. Large number of applications either don't expect exceptions from the within, or catch very specific exception types at appropriate code boundaries. Even those that have such catch all handlers, might have it at a deeper code boundary, not at the topmost level in Main method. If you are more comfortable having a catch all handler anywhere inside your code, you should add a source suppression for CA1031 around it, which also acts as a nice documentation point for your reasoning behind having the handler there.
